### PR TITLE
feat: filtering and sorting for questions repo

### PR DESCRIPTION
**Pull request by firwer**
Features:
1. FIltering by topics and question difficulty
2. Sorting of table values
image
image

### CODE REVIEW

**Comment by javinchua** [comment]
hey noticed u are implementing ur own functionality for the sorting and filtering but would it be easier to use the [<DataTable>](https://ui.shadcn.com/docs/components/data-table) component from `shacn/ui` instead? 

It will help to keep the UI consistent and also they provide lots of the functionalities which we would need such as:
- Sorting
- Filtering
- Pagination
- And other features if we want to use it in the future.

You can reference this [example](https://ui.shadcn.com/examples/tasks) here to see how the <DataTable> would look like with shadcn
This is the [source code](https://github.com/shadcn-ui/ui/tree/main/apps/www/app/(app)/examples/tasks/components) for the example above. 
I think all the components they have here will probably be useful to use except for:
- `[user-nav.tsx](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/user-nav.tsx)`
- [data-table-view-options.tsx](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-view-options.tsx)

The <DataTable> uses [@tanstack/react-table](https://tanstack.com/table/latest/docs/introduction) under the hood fyi. 
If your unsure on how to do it i wouldnt mind helping u with it.

**Comment by javinchua** [comment]
Also, correct me if im wrong but i think for the toast we should import `useToast` instead and use it as a hook, rather than directly import `toast` as i think `toast` is used if you are not using it within a react component.
```
import { useToast } from "@hooks/use-toast"
const { toast } = useToast()
toast({
...
})
```

**Comment by jonyeokj** [comment]
The current filtering logic uses an `OR` condition for multiple categories, returning entries that match any selected category. I believe that it is more intuitive to use an `AND` condition to find entries that match all selected categories.

To add on to Javin's point on UI consistency, you can check on [this multi-select component](https://github.com/sersavan/shadcn-multi-select-component). If you can incorporate it, you can update the create and edit modals with multi-select as well.

**Comment by firwer** [comment]
> hey noticed u are implementing ur own functionality for the sorting and filtering but would it be easier to use the [DataTable](https://ui.shadcn.com/docs/components/data-table) component from `shacn/ui` instead?
> 
> It will help to keep the UI consistent and also they provide lots of the functionalities which we would need such as:
> 
> * Sorting
> * Filtering
> * Pagination
> * And other features if we want to use it in the future.
> 
> You can reference this [example](https://ui.shadcn.com/examples/tasks) here to see how the DataTable would look like with shadcn This is the [source code](https://github.com/shadcn-ui/ui/tree/main/apps/www/app/(app)/examples/tasks/components) for the example above. I think all the components they have here will probably be useful to use except for:
> 
> * [user-nav.tsx](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/user-nav.tsx)
> * [data-table-view-options.tsx](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-view-options.tsx)
> 
> The DataTable uses [@tanstack/react-table](https://tanstack.com/table/latest/docs/introduction) under the hood fyi. If your unsure on how to do it i wouldnt mind helping u with it.

Got it, i'll try this out. Might have been confused with your original question about using datatable. will be looking into this, thanks!

**Comment by javinchua** [comment]
Hey just took a quick look at the PR. I think its better but some points i would have are:
- You could reference [example](https://github.com/shadcn-ui/ui/tree/main/apps/www/app/(app)/examples/tasks/components) more to create the components and also abstract them out. currently its all in one place so its not super readable.
- Correct me if im wrong but it also seems like you are using the @tanstack/react-table library instead of the shadcn's [DataTable](https://ui.shadcn.com/docs/components/data-table) component.
I would recommend following these steps to make it easier:
1. Run the following command: `pnpm dlx shadcn@latest add table` => This should create a `DataTable` component in @/components/ui
2. Follow this to add more of the other components which adds on to the functionality of DataTable. This includes the following:
- [Column Headers](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-column-header.tsx). This basically defines the functionalities of the column headers in the table into an abstracted format.
- [Pagination](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-pagination.tsx) Defines what we see at the footer of the table, including next page, number of rows and more
- [Row actions](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-row-actions.tsx) Optional for now but will be allowing us to choose the actions we can perform on a particular row(e.g. edit, delete)
- [Filter](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-faceted-filter.tsx). This basically helps define the filter functionalities.
- [Toolbar](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table-toolbar.tsx) Utilises the `Filter` component above to implement our filter functionality. Also comes up with a search function
3. Using these components we can import them within our `DataTable` component that was generated by shadcn previously, similar to what this [example](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/data-table.tsx) is doing.
4. [Columns](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/components/columns.tsx). Where we define all the different columns we have, whether they will have a filter/sort functionality, as well as define the filter functions. 
5. We import our `data` and `Columns` and pass it into `DataTable` in the questions page. See [example](https://github.com/shadcn-ui/ui/blob/main/apps/www/app/(app)/examples/tasks/page.tsx#L61)

Also, dont remove the `Suspense` because we are using `useSuspenseQuery`. Basically, `useSuspenseQuery` will call the api and cause the component it is in to `Suspend` until the api returns something. While it is `suspended` it triggers the nearest `Suspense` component that displays the fallback (Usually its a `Skeleton` which symbolises the component is loading). If we remove the `Suspense` the effect will propagate up till the root where our next `Suspense` boundary is which isnt ideal cause the fallback will be different and not accurate. Or worse, if theres no Suspense boundary then there isnt any more fallback.

Also, i think u can try merging in main into your code first to solve the `toast` stuff.

**Comment by firwer** [comment]
> 1. pnpm dlx shadcn@latest add table
Just one point on your remark, 

I am not exactly sure what do you mean by it will create a <DataTable/> object because the component added is <Table> not   <DataTable>as you have pointed out. Not sure if I misunderstood something here. DataTable is not natively part of the shadcn.

Perhaps I think you mean something like this https://ui.shadcn.com/docs/components/data-table just that abstract it out into a custom component called <DataTable> correct? (like in the demo url from shadcn). Just checking before I start on anything.

Yup, good point, for the rest, I will abstract it out where possible to tidy things up

**Comment by firwer** [comment]
> > 1. pnpm dlx shadcn@latest add table
> >    Just one point on your remark,
> 
> I am not exactly sure what do you mean by it will create a `<DataTable/>` object because the component added is `<Table>` not `<DataTable>`as you have pointed out. Not sure if I misunderstood something here. DataTable is not natively part of the shadcn.
> 
> Perhaps I think you mean something like this https://ui.shadcn.com/docs/components/data-table just create a custom component called `<DataTable>` based on `<Table>` correct? (like in the demo url from shadcn). Just confirming we are on the same page before I start on anything.
> 
> Yup, got it, for the rest, I will abstract it out where possible to tidy things up

o

**Comment by firwer** [comment]
> > 1. pnpm dlx shadcn@latest add table
> >    Just one point on your remark,
> 
> I am not exactly sure what do you mean by it will create a `DataTable` component because the component added is `<Table>` not `<DataTable>`as you have pointed out. Not sure if I misunderstood something here. DataTable is not natively part of the shadcn.
> 
> Perhaps I think you mean something like this https://ui.shadcn.com/docs/components/data-table just create a custom component called `<DataTable>` based on `<Table>` correct? (like in the demo url from shadcn). Just confirming that I am on the same page
> 
> Yup, got it, for the rest, I will abstract it out where possible to tidy things up

update: refer to the latest commit, let me know if this is something you have in mind

**Comment by javinchua** [comment]
image
(putting this image for reference)

Yup i think looks a lot better and a good first working version. Do you mind changing the filtering logic to be using `AND` rather than `OR` as mentioned above in jon's comment? 

### List of future improvements (don't have to address them in this PR, I can help to implement these too)
- When toggling the sorting of a column, alternate between `ArrowUp` and `ArrowDown` to show the sorted order
- Filtering UI doesn't look super clean
ours:
 image
suggested: 
image
- Pagination lacks page numbers and allowing users to choose how many rows they want
- Can consider between FE pagination and BE pagination
- Searching of questions by name

File: project/apps/web/app/questions/components/question-table/columns.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+// Filtering functions
+const complexityFilter = (row: any, columnId: string, filterValue: string[]) => {
+  return filterValue.length === 0 || filterValue.includes(row.getValue(columnId));
+};
+
+const categoryFilter = (row: any, columnId: string, filterValue: string[]) => {
+  const rowValues = row.getValue(columnId) as string[];
+  return (
+    filterValue.length === 0 ||
+    filterValue.some((value: string) => rowValues.includes(value))
```
Comment (javinchua) [inline project/apps/web/app/questions/components/question-table/columns.tsx:35]:
```suggestion
    filterValue.every((value: string) => rowValues.includes(value))
```

**Comment by javinchua** [comment]
> I am not exactly sure what do you mean by it will create a DataTable component because the component added is <Table> not <DataTable>as you have pointed out. Not sure if I misunderstood something here. DataTable is not natively part of the shadcn.
Oops ps, i thought they automatically created it for u haha

**Comment by javinchua** [comment]
Oh and just another comment, i dont know how useful sorting based on `Categories` is, given that `Categories` is an array. Seems like we are currently sorting based on the first element of the array, but this way of sorting isn't useful because our order of elements within the array is arbitrary.
Only useful way of sorting I can think of is by length of array.
What do you guys think? 
@jeffrey-jian @jonyeokj @firwer @gongg21

**Comment by firwer** [comment]
image

The latest changes should resolve the filter related issues (i.e. categories). In addition, I have

> Oh and just another comment, i dont know how useful sorting based on `Categories` is, given that `Categories` is an array. Seems like we are currently sorting based on the first element of the array, but this way of sorting isn't useful because our order of elements within the array is arbitrary. Only useful way of sorting I can think of is by length of array. What do you guys think? @jeffrey-jian @jonyeokj @firwer @gongg21

image

I tested the sorting of category by numbers and it appears to be more intuitive, guess we can stick to this.

The latest changes should resolve the filter related issues (i.e. categories). In addition, I have added a search input as well as tweak the icon for table sort. It should address most of the more pressing points you highlighted earlier.

As for the remaining features mentioned like pagination imho i think it would be a good to have but not high priority as of yet so perhaps we can maybe create a kanban and add them in to keep track for future progress like you mentioned above.

**Review by javinchua — APPROVED** [review]
lgtm

**Pull request merged**